components/org.wso2.carbon.crypto.impl/src/main/java/org/wso2/carbon/crypto/impl/DefaultCryptoService.java

```diff
@@ -1006,4 +1065,50 @@ private void failIfHybridDecryptOperationInputsAreInvalid(HybridEncryptionOutput
             throw new CryptoException(errorMessage);
         }
     }
+
+    private void failIfInternalCryptoInputsAreNotValid(byte[] data, String algorithm, String internalCryptoProviderType,
+                                                       String operation) throws CryptoException {
+
+        if (data == null) {
+            throw new CryptoException(String.format("Content provided for the %s operation can't be null", operation));
+        }
+
+        if (StringUtils.isBlank(algorithm)) {
+            throw new CryptoException("Algorithm can't be empty");
+        }
+
+        if (StringUtils.isBlank(internalCryptoProviderType)) {
+            throw new CryptoException("Internal Crypto Provider type can't be empty");
+        }
+    }
+
+    /**
+     * Returns the most suitable {@link InternalCryptoProvider} for the operations, among the registered providers.
+     *
+     * @param internalCryptoProviderType preferred internal crypto provider.
+     * @return most suitable internal provider for the usecase.
+     * @throws CryptoException
+     */
+    private InternalCryptoProvider getMostSuitableInternalProvider(String internalCryptoProviderType)
+            throws CryptoException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Looking for the most suitable internal crypto provider.");
+        }
+
+        if (StringUtils.isNotBlank(internalCryptoProviderType)) {
+
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Preferred internal crypto provider '%s' received from method api.",
```

Comment:
Should we reflect as method api. Shall we say 'from registered providers'

Reply:
Fixed with 04d2d03f3116052f1e52061e0a3744f894622eff